dapr-ext-langgraph/dapr/ext/langgraph/dapr_checkpointer.py

```diff
@@ -0,0 +1,158 @@
+import json
+from typing import Any, Sequence, Tuple
+
+from langchain_core.load import dumps
+from langchain_core.runnables import RunnableConfig
+
+from dapr.clients import DaprClient
+from langgraph.checkpoint.base import BaseCheckpointSaver, Checkpoint, CheckpointTuple
+
+
+class DaprCheckpointer(BaseCheckpointSaver[Checkpoint]):
+    """
+    Dapr-backed LangGraph Checkpointer that persists checkpoints to a Dapr state store.
+    Compatible with LangGraph >= 0.3.6 and LangChain Core >= 1.0.0.
+    """
+
+    REGISTRY_KEY = 'dapr_checkpoint_registry'
+
+    def __init__(self, store_name: str, key_prefix: str):
+        self.store_name = store_name
+        self.key_prefix = key_prefix
+
+    # helper: construct Dapr key for a thread
+    def _get_key(self, config: RunnableConfig) -> str:
+        thread_id = None
+        if isinstance(config, dict):
+            thread_id = config.get('configurable', {}).get('thread_id')
+            if not thread_id:
+                thread_id = config.get('thread_id')
+        if not thread_id:
+            thread_id = 'default'
+        return f'{self.key_prefix}:{thread_id}'
+
+    # restore a checkpoint
+    def get_tuple(self, config: RunnableConfig) -> CheckpointTuple | None:
+        key = self._get_key(config)
+        with DaprClient() as client:
```

Comment:
it's better to create one single instance of the client in the `__init__` instead of a new instance every time.